src/sentry/sentry_metrics/multiprocess.py

```diff
@@ -536,12 +537,31 @@ def __init__(
         self.__producer_topic = settings.KAFKA_TOPICS[settings.KAFKA_SNUBA_METRICS].get(
             "topic", "snuba-metrics"
         )
+        self.__commit_function = commit_function
 
         self.__closed = False
         self.__metrics = get_metrics()
+        self.__produced_message_offsets = []
+        self.__callbacks = 0
 
     def poll(self) -> None:
-        pass
+        if self.__callbacks and self.__produced_message_offsets:
+            success_messages = self.__produced_message_offsets[: self.__callbacks]
+            latest_offset_by_partition = {}
+            for message in success_messages:
+                latest = latest_offset_by_partition.get(message.partition)
+
+                if latest and message.position.offset < latest.offset:
+                    raise OutOfOrderOffset()
+
+                latest_offset_by_partition[message.partition] = message.position
```

Comment:
I think there is a simpler way to do this that does not require you to keep a list of all the partitions and offsets.
You can get partition and offset passed to the callback by making the callback a partial function:

```
def callback(self, partition: Partition, offset:Position, error: Any, message: Any) -> None:
     # do whatever you want with the offset, which basically means something like:
     self.__offsets[partition] = offset
     # also decrement the counter
```

Now, in order to use the callback this way you need to pass a apartial function to the producer.

```
self.__producer.produce(
            topic=self.__producer_topic,
            key=None,
            value=message.payload.value,
            on_delivery=partial(self.callback, message.partition, message.position.offset)
        )
      # also increment a counter
```

Then in this method you can just do this:
- If the difference between what was produced and what was acknowledged is greater than a value, start with 90,000, Do a long poll .
- Every X messages acknowledged (let's do 30,000), or every Y seconds commit

Please keep the following in runtime flags that we can change without restarting as we will do quite some tuning.
- the threshold after which you do the long poll 
- the batch size for commit 
- the batch time for commit 
- the long poll timeout

Obviously you cannot reload the the flag at each iteration or you will take flagr down. 
You need to reload these parameter every few seconds.